plugins/emoji/plugin.js

```diff
@@ -0,0 +1,136 @@
+			// jscs:enable
+
+			if ( !editor._ ) {
+				editor._ = {};
+			}
+
+			if ( !editor._.emojiList ) {
+				editor._.emojiList = defaultEmojiList;
+			}
+			var emoji = editor._.emojiList,
+				forbiddenScope = editor.config.emoji_preventTransformation || [ 'pre', 'code' ],
+				charactersToStart = editor.config.emoji_numberOfCharactersToStart || 0;
+
+			editor.on( 'instanceReady', function() {
+				var emoji = new CKEDITOR.plugins.autocomplete( editor, getTextTestCallback(), dataCallback );
+
+				emoji.view.itemTemplate = new CKEDITOR.template( '<li data-id="{id}" style="overflow:hidden;text-overflow:ellipsis;white-space:nowrap;">{symbol} {id}</li>' );
+				emoji.getHtmlToInsert = function( item ) {
+					return item.symbol;
+				};
+
+			} );
+
+			editor.on( 'toHtml', function( evt ) {
+				var filter = new CKEDITOR.htmlParser.filter( {
+					text: function( value, element ) {
+						var notConverted = element.getAscendant( getConfigForNotConvertedTags() );
+						if ( notConverted ) {
+							return value;
+						}
+						var hits = value.match( /:[a-zA-Z_-]+?:/g );
+						if ( hits ) {
+							for ( var i = 0; i < hits.length; i++ ) {
+								var item = emoji.find( function( element ) {
+									return element.id === hits[ i ];
+								} );
+								if ( item ) {
+									value = value.replace( item.id, item.symbol );
+								}
+							}
+						}
+						return value;
+					}
+				} );
+				filter.applyTo( evt.data.dataValue );
+			} );
+
+			function getTextTestCallback() {
+				return function( range ) {
+					if ( !range.collapsed || range.startContainer.getAscendant( getConfigForNotConvertedTags() ) ) {
+						return null;
+					}
+					return CKEDITOR.plugins.textMatch.match( range, matchCallback );
+				};
+			}
+
+			function matchCallback( text, offset ) {
+				var left = text.slice( 0, offset ),
+					match = left.match( new RegExp( ':\\S{' + charactersToStart + '}\\S*$' ) );
+
+				if ( !match ) {
+					return null;
+				}
+
+				return { start: match.index, end: offset };
+			}
+
+			function dataCallback( query, range, callback ) {
+				var data = CKEDITOR.tools.array.filter( emoji, function( item ) {
+					return item.id.indexOf( query.slice( 1 ) ) !== -1;
+				} ).sort( function( a, b ) {
+					// Sort at the beginning emoji starts with given query.
+					var emojiName = query.substr( 1 ),
+						isAStartWithEmojiName = a.id.substr( 1, emojiName.length ) === emojiName,
+						isBStartWithEmojiName = b.id.substr( 1, emojiName.length ) === emojiName;
+
+					if ( isAStartWithEmojiName && isBStartWithEmojiName || !isAStartWithEmojiName && !isBStartWithEmojiName ) {
+						return a.id === b.id ? 0 : ( a.id > b.id ? 1 : -1 );
+					} else if ( isAStartWithEmojiName ) {
+						return -1;
+					} else {
+						return 1;
+					}
+				} );
+				callback( data );
+			}
+
+			function getConfigForNotConvertedTags() {
```

Comment:
I've used name `hasForbiddenParent` as it more appropriate to this solution :)

Reply:
I don't think that it's helping much.

In fact the oposite, you're bounding function name with the implementation of it. Now what will you do if there will be a requirement to disable mentioning for classic editors?

1. You add an `editor` instance as an argument (because you need to use info from the editor inside the function).
1. Your proposed function name no longer matches the purpose (is misleading). So are you going to rename it into `hasForbiddenParentAndIsInNonClassicalEditor`?

`isAllowed` or `isEmojiAllowed` does not have this issue.

Reply:
I've made improvement on different scope of a problem. Removing emoji in classic will be easily to extend. Take a look:
```
var preventEmojiConversion = element.getAscendant( hasForbiddenParent );
if ( preventEmojiConversion ) {
	return value;
}
```
I can rename `preventEmojiConversion` to `isEmojiAllowed` if you think this will be more suitable here.
Also it is easy to extend this with classical editor like:
```
var preventEmojiConversion = element.getAscendant( hasForbiddenParent ) || isNonClassicalEditor();
```

Unfortunately `getAscendant`method break after first true match. So in case like `p > strong > em > a`. And configuration where someone prevent emoji in strong, but current selection is in anchor. Anchor would be detected as allowed parent (it is not on forbidden list), loop will break and emoji would appear.
I need to search all ascendants for forbidden elements and break loop on forbidden element and in such case don't activate emoji conversion.